File: FvU3D.R



##Adjust working directory to the location of genotypes.csv and gametes.csv
setwd('/home/hannes/Dropbox/scripts/FvsU/')




#defines the dimensions of the genotype array

initialise<-function(wi=10, len=40, popsize=40){
        genotypenr<-11
        genotypenames<-c('Ubb', 'Fa', 'Uab', 'Fb', 'Uaa', 'UUbb', 'FF', 'FUb', 'UUab', 'FUa', 'UUaa')
        a<-array(0, dim=c(wi, len, genotypenr),
                dimnames=list(paste0('w', 1:wi), paste0('l', 1: len), genotypenames))
        
        abu<-matrix(popsize, wi, len)
        
        a[,1:20, c(1,6)]<-0.5
        a[,21:40, c(2,7)]<-0.5
        for(i in 1:dim(a)[3]){
                a[,,i]<-a[,,i]*abu
        }
        
        
        
        return(list(g=a, c=0, left=a[,1,], right=a[,len,], to=a[1,,], bo=a[wi,,], abu=abu))
}

#genotypes<-initialise()
#genotypes$g
#genotypes$c
#genotypes$left
#genotypes$right
#genotypes$to
#genotypes$bo
#genotypes$abu

#Selection parameters (cf. Veltsos et al. 2008 Tab. 1)

sm<-0.05         #selection against b in males
dm<-1           #dominance of a in males
sf<-0.02           #selection against a in females
df<-1           #dominance of a in females
sh<-0           #"heterozygote disadvantage"
sd<-0           #"dosage compensation"
sc<-0           #"sex chromosome coadaption"

disp<-0.03      #dispersal in each direction

#vector of selection acting on genotypes (SELECTion vectOR)
selector<-c(1-sm, 1, (1-sm*dm)*(1-.5*sd)*(1-.5*sc), (1-sm)*(1-.5*sd)*(1-sc), (1-.5*sd)*(1-sc),
            1, 1, (1-sh)*(1-.5*sd)*(1-.5*sc), (1-sf*df)*(1-.5*sd)*(1-.5*sc), (1-sf*df)*(1-sh)*(1-.5*sc), (1-sf)*(1-sd)*(1-sc))
names(selector)<-c('Ubb', 'Fa', 'Uab', 'Fb', 'Uaa', 'UUbb', 'FF', 'FUb', 'UUab', 'FUa', 'UUaa')

selector


##What gamete ratios (rows) result from which genotype (columns)? This table will tell us:
##ajust path to file
gamconv<-as.matrix(read.table('gametes.csv', sep='\t', row.names=1, head=T))
gamconv

#What genotype results from which combination of gametes?
##ajust path to file
genconv<-as.matrix(read.table('genotypes.csv', sep='\t', row.names=1, head=T))
genconv


#'zygotes from gametes' creates n zygotes from a population's gamete vector x
zfg<-function(x,n){             
        #male gemetes are x[1:5]
        mametes<-sapply(runif(n, 0, sum(x[1:5])), function(y) min(which(y<cumsum(x[1:5]))))
        #female gametes are in x[6:8]
        fametes<-sapply(runif(n, 0, sum(x[6:8])), function(z) min(which(z<cumsum(x[6:8]))))
        #find zygote genotype in imported table
        zygotes<-genconv[cbind(mametes, fametes)]
        zygotetab<-vector('numeric',11)
        names(zygotetab)<-c('Ubb', 'Fa', 'Uab', 'Fb', 'Uaa','UUbb', 'FF', 'FUb', 'UUab', 'FUa', 'UUaa')
        for(v in zygotes){zygotetab[which(names(zygotetab)==v)]<-zygotetab[which(names(zygotetab)==v)]+1}
        return(zygotetab)
}

##########
##########Find a way to visualise the results!

Fratio<-function(genot=genotypes$g){
        apply(genot, c(1,2), function(x) sum(x*c(0,1,0,1,0, 0,2,1,0,1,0))/sum((x * c(1,1,1,1,1, 2,2,2,2,2,2))))
}

YFratio<-function(genot=genotypes$g){
        apply(genot, c(1,2), function(x) sum(x*c(0,1,1,0,2, 0,0,0,1,1,2))/sum((x * c(2,1,2,1,2, 2,0,1,2,1,2))))
}


#####Print results to png
print_to_png<-function(c, data){
        
        if(c %% 100 ==0){
                
                png(paste('sm', sm, 'dm', dm, 'sf', sf, 'df', df, 'sh', sh, 'sd', sd, 'sc', sc, 'disp', disp,
                          sprintf('gen_%05d.png', c), sep="_"), width=1400, height=800)
                
                par(mfrow=c(1,2))
                print(dim(data))
                image(Fratio(data), asp=dim(data)[2]/dim(data)[1], frame=F, col=heat.colors(100), breaks=seq(0,1,0.01), axes=F, main=paste('F,', c, 'generations'))
                #image(abc$g[,,6], col=heat.colors(40), breaks=seq(0,40,1), main=paste(abc$c, 'generations'), asp=4, frame=F)
                contour(Fratio(data), add=T, levels=c(0, .05, .5, .95,1))
                abline(h=0.5, lty=2)
                image(YFratio(data), asp=dim(data)[2]/dim(data)[1], frame=F, col=heat.colors(100), breaks=seq(0,1,0.01), axes=F, main=paste('Y|not F,', c, 'generations'))
                contour(YFratio(data), add=T, levels=c(0, .05, .5, .95,1))
                abline(h=0.5, lty=2)
                
                dev.off()
                               
        }
}


#########
#########The new TIME GOES BY function
tgb<-function(genot=genotypes, gens=10, open=F, printout=F){
        
        #defining and internal object representing the genotypes
        
        dimensions<-dim(genot$g)
        #print(dimensions)
        intgen<-array(0, c(dimensions[1]+2,dimensions[2]+2,dimensions[3]))
        
        intgen[2:(dimensions[1]+1), 2:(dimensions[2]+1),]<-
                intgen[2:(dimensions[1]+1), 2:(dimensions[2]+1),]+genot$g
        
        #if it is closed, create a matrix, closedfact
        if(open==F){
                
                f4<-1-4*disp
                f3<-1-3*disp
                f2<-1-2*disp
                
                closedfact<-array(0, dimensions)
                closedfact[2:(dimensions[1]-1), 2:(dimensions[2]-1),]<-f4
                closedfact[1:(dimensions[1]), c(1, dimensions[2]),]<-f3
                closedfact[c(1, dimensions[1]), 1:(dimensions[2]),]<-f3
                closedfact[1,1,1:dimensions[3]]<-f2
                closedfact[dimensions[1],1,1:dimensions[3]]<-f2
                closedfact[1,dimensions[2],1:dimensions[3]]<-f2
                closedfact[dimensions[1],dimensions[2],1:dimensions[3]]<-f2
                #print(closedfact)
                
                #if it is not closed, intgen's margins will equal their neighbouring elements        
        } else {               
                
                intgen[2:(dimensions[1]+1),   1,]<-genot$left
                intgen[2:(dimensions[1]+1), dimensions[2]+2,]<-genot$right
                intgen[1,  2:(dimensions[2]+1),]<-genot$to
                intgen[dimensions[1]+2, 2:(dimensions[2]+1),]<-genot$bo
                
        }
        #for debug        
        #        return(intgen)
        #}       
        #
        #tgb(open=F)
        #abc
        #abc<-tgb()        
        #genotypes$left
        #abc[2:11,1,]<-genotypes$left
        #abc[,1,]
        #1     
        
        
        #the actual loop starts here
        if(open==F){
                for(i in 1:gens){
                        
                        #dispersal
                        intgen[2:(dimensions[1]+1), 2:(dimensions[2]+1),]<-
                                intgen[2:(dimensions[1]+1), 2:(dimensions[2]+1),]*closedfact+
                                intgen[1:(dimensions[1]),   2:(dimensions[2]+1),]*disp+
                                intgen[3:(dimensions[1]+2), 2:(dimensions[2]+1),]*disp+
                                intgen[2:(dimensions[1]+1), 1:(dimensions[2]),]*disp+
                                intgen[2:(dimensions[1]+1), 3:(dimensions[2]+2),]*disp
                        
                        #gametes
                        #get array of gamete ratios
                        gametes<-apply(intgen[2:(dimensions[1]+1), 2:(dimensions[2]+1),], c(1,2),
                                       function(x) rowSums(matrix(x*selector,8,11,byrow=T)*gamconv))                        
                        
                        #get matrix of genotypes (involves drift by samplig from a 'uniform' distribution)
                        
                        for(k in 1:dimensions[1]){
                                #print(k)
                                for(l in 1:dimensions[2]){
                                        #print(l)
                                        intgen[k+1, l+1,]<-zfg(gametes[,k,l], genot$abu[k,l])        
                                }
                        }
                        
                        
                        
                        
                        #counter
                        genot$c<-genot$c+1
                        print(paste(genot$c, 'generations'))
                        
                        if(printout==T){print_to_png(genot$c, intgen[2:(dimensions[1]+1),2:(dimensions[2]+1),])}
                        
                        
                }
        } else {
                for(i in 1:gens){
                        
                        #dispersal
                        intgen[2:(dimensions[1]+1), 2:(dimensions[2]+1),]<-
                                intgen[2:(dimensions[1]+1), 2:(dimensions[2]+1),]*(1-4*disp)+
                                intgen[1:(dimensions[1]), 2:(dimensions[2]+1),]*disp+
                                intgen[3:(dimensions[1]+2), 2:(dimensions[2]+1),]*disp+
                                intgen[2:(dimensions[1]+1), 1:(dimensions[2]),]*disp+
                                intgen[2:(dimensions[1]+1), 3:(dimensions[2]+2),]*disp
                        
                        
                        
                        #gametes
                        #get array of gamete ratios
                        gametes<-apply(intgen[2:(dimensions[1]+1), 2:(dimensions[2]+1),], c(1,2),
                                       function(x) rowSums(matrix(x*selector,8,11,byrow=T)*gamconv))                        
                        
                        #get matrix of genotypes (involves drift by samplig from a 'uniform' distribution)
                        
                        for(k in 1:dimensions[1]){
                                #print(k)
                                for(l in 1:dimensions[2]){
                                        #print(l)
                                        intgen[k+1, l+1,]<-zfg(gametes[,k,l], genot$abu[k,l])        
                                }
                        }
                        
                        
                        
                        
                        #counter
                        genot$c<-genot$c+1
                        print(paste(genot$c, 'generations'))
                        if(printout==T){print_to_png(genot$c, intgen[2:(dimensions[1]+1),2:(dimensions[2]+1),])}
                        
                }
        }
        
        
        
        return(list(g=intgen[2:(dimensions[1]+1), 2:(dimensions[2]+1),], c=genot$c, left=genot$left, right=genot$right,
                    to=genot$to, bo=genot$bo, abu=genot$abu))
        
}







genotypes<-initialise(40,40,40)





#running tgb in a loop plotting graphs every 10 generatins
for (i in 1:10000){
        genotypes<-tgb(genotypes,gens=10,open=F)
        par(mfrow=c(1,2))
        image(Fratio(genotypes$g), asp=dim(genotypes$g)[2]/dim(genotypes$g)[1], frame=F, col=heat.colors(100), breaks=seq(0,1,0.01), axes=F, main=paste('F,', genotypes$c, 'generations'))
        contour(Fratio(genotypes$g), add=T, levels=c(0, .05, .5, .95,1))
        abline(h=0.5, lty=2)
        
        image(YFratio(genotypes$g), asp=dim(genotypes$g)[2]/dim(genotypes$g)[1], frame=F, col=heat.colors(100), breaks=seq(0,1,0.01), axes=F, main=paste('Y|not F,', genotypes$c, 'generations'))
        contour(YFratio(genotypes$g), add=T, levels=c(0, .05, .5, .95,1))
        abline(h=0.5, lty=2)
}


#tbg can be run in printout mode saving heat map and contour every 100 generations
#files will be saved to the working directory and will be names with the parameters chosen
tgb(genotypes, gens=20000, printout=T)



######################
######################
######################
##TEST AREA


#100 gens need approx. 22 s on hardy

File: FvU3D.py
#!/usr/bin/python


#FvU3D.py currently takes three pramameters all of which have to be supplied when calling the programme.
#These are sm (selection against the)
#EXAMPLE: python FvU3D.py .05 1 .1 1 0 0 0 .03 0 1



import numpy as np
import sys
import pickle
import gzip
import statsmodels.api as stm
#import time



#The 11 genotypes are: 'Ubb', 'Fa', 'Uab', 'Fb', 'Uaa', 'UUbb', 'FF', 'FUb', 'UUab', 'FUa', and 'UUaa' (0-10)
#The 8 gametes (5 male + 3 female) are: 'mF', 'ma', 'mb', 'mUa', 'mUb', 'fF', 'fUa', and 'fUb' (0-7) 

#number of Fs containd in genotypes
Fgenotypes = np.array([0,1,0,1,0, 0,2,1,0,1,0])
#number of potential Fs (depending on sex)
Fpot = np.array([1,1,1,1,1, 2,2,2,2,2,2])
Fabs = np.array([1,0,1,0,1, 2,0,1,2,1,2])
#number of Ys containd in genotypes
Ygenotypes = np.array([0,1,1,0,2, 0,0,0,1,1,2])
#number of potential Ys (depending on wheter and how many Fs are present)
Ypot = np.array([2,1,2,1,2, 2,0,1,2,1,2])
Yabs = np.array([2,0,1,1,0, 2,0,1,1,0,0])

def Fratio(x):
	return np.sum(x*Fgenotypes, axis=2)/np.sum(x*Fpot, axis=2)



def YFratio(x):
	return np.sum(x*Ygenotypes, axis=2)/np.sum(x*Ypot, axis=2)


def Fhits(x):
	return np.sum(x*Fgenotypes, axis=2)

def Fmisses(x):
	return np.sum(x*Fabs, axis=2)

def Yhits(x):
	return np.sum(x*Ygenotypes, axis=2)

def Ymisses(x):
	return np.sum(x*Yabs, axis=2)

genconv=np.array([6,9,7,1,4,2,3,2,0,9,10,8,7,8,5]).reshape(5,3)

gamconv=np.array([0, 2, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 2, 1, 0, 2, 0, 0, 0,\
	0, 0, 0, 2, 0, 1, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 2, 0, 0, 0,\
	0, 0, 0, 2, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 4, 2,\
	0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 2, 4, 0, 0, 0, 0, 0, 4, 0, 2, 2, 0, 0]).reshape(8,11)


def params():
	
	if len(sys.argv) <> 11:
		print 'Wrong number of parameters supplied. Exit.'
		exit()

	global 	sm, dm, sf, df, sh, sd, sc, disp, grad, run

	sm = float(sys.argv[1])
	dm = float(sys.argv[2])
	sf = float(sys.argv[3])
	df = float(sys.argv[4])
	sh = float(sys.argv[5])
	sd = float(sys.argv[6])
	sc = float(sys.argv[7])
	
	disp = float(sys.argv[8])
	grad = int(sys.argv[9])
	run = int(sys.argv[10])


	#vector of selection acting on genotypes (SELECTion vectOR)
	global selector
	selector=np.array([[1-sm, 1, (1-sm*dm)*(1-.5*sd)*(1-.5*sc), (1-sm)*(1-.5*sd)*(1-sc),\
		(1-.5*sd)*(1-sc), 1, 1, (1-sh)*(1-.5*sd)*(1-.5*sc), (1-sf*df)*(1-.5*sd)*(1-.5*sc),\
		(1-sf*df)*(1-sh)*(1-.5*sc), (1-sf)*(1-sd)*(1-sc)]])


#Intialises a 3D genotype array ['g']
#(plus a counter['c'], an abundance matrix ['abu'] and other stuff
#that may become important later).
#Returns a dictionary containing all that stuff
def initialise(wi=10, leng=120, popsize=40):

	a=np.zeros([wi, leng, 11])

	abu=np.repeat(popsize , wi*leng).reshape(wi, leng)


	a[:, 0:20, [1,6]]=.5	#fused
	a[:, 20:, [0,5]]=.5	#unfused

	a=a*abu[:,:,np.newaxis]

	c=0

	trans=np.array([0,0])

	print "Matrix initialised with %dx%d populations of %d individuals each." % (wi, leng, popsize)
	return {'g': a, 'abu': abu, 'c': c, 'fm': a[:,[0],:], 'um': a[:,[leng-1],:], 'trans':trans}




#Get Gamete Ratios, takes one 1D np array (genotype vector) as argument and
#returns a 1D 8-item gamete vector
def ggr(x):	

	return np.sum(x*gamconv, axis=1)

#GAmete SAample
#samples gametes according to gamete ratios, does sex and
#returns the new generations genotypes
def gasa(x):	

	#Male gAMETES
	mametes=np.repeat(np.arange(5),np.random.multinomial(40, x[0:5]/sum(x[0:5]), 1).flatten())
	np.random.shuffle(mametes)

	#Female gAMETES
	fametes=np.repeat(np.arange(3),np.random.multinomial(40, x[5:8]/sum(x[5:8]), 1).flatten())
	np.random.shuffle(fametes)

	##FASTAER SEX NOW - MULTIPLE SELECTION FROM ARRAY BY "ADVANCED" INDEXING
	a = np.bincount(genconv[mametes, fametes])
	#topping upt the array in case not all 11 genotypes were produced
	a = np.hstack((a, np.zeros(11-a.shape[0])))
	return a

#the heart of this programme, the Time Goes By function
def tgb(genot, gens=10):

	#dimensions of the genotype matrix	
	dimensions=np.array(genot['g'].shape)
	
	
	#intgen is the tbg-internal version of the genotype array
	#it is two elements wirde in the firs two dimensions to allow
	#distribution to be modelled by addition of an offset vrsion of the array
	
	intgen=np.zeros(dimensions+np.array([2,2,0]))
	
	intgen[1:(dimensions[0]+1), 1:(dimensions[1]+1), :]=genot['g']

	#DISPersal FACTor is an arry with the same dimensions as the genotype array
	#it is being multiplid with the genotypes to account for margin effects
#	dispfact=np.repeat(1-disp*4, np.prod(dimensions)).reshape(dimensions)
#	dispfact[[0,dimensions[0]-1],:,:]=1-disp*3
#	dispfact[:,[0,dimensions[1]-1],:]=1-disp*3
#	dispfact[[0, 0, dimensions[0]-1, dimensions[0]-1],[0, dimensions[1]-1, 0, dimensions[1]-1],:]=1-disp*2
#	print dispfact

	for i in range(gens):
		#dispersal
		intgen[1:(dimensions[0]+1), 1:(dimensions[1]+1), :]=\
			intgen[1:(dimensions[0]+1), 1:(dimensions[1]+1), :]*(1-4*disp)+\
			intgen[0:(dimensions[0]), 1:(dimensions[1]+1), :]*disp+\
			intgen[2:(dimensions[0]+2), 1:(dimensions[1]+1), :]*disp+\
			intgen[1:(dimensions[0]+1), 0:(dimensions[1]), :]*disp+\
			intgen[1:(dimensions[0]+1), 2:(dimensions[1]+2), :]*disp
		#print intgen[:,18:22,:]
		

		#selection acts
		#(selection vector is being multiplied with the genotype array)
		gametes=intgen[1:(dimensions[0]+1), 1:(dimensions[1]+1), :]*selector


		#gamete ratios, from the genotypes arry every vector along axis 2 (= genotype axis)
		#is being multiplied with the genotype conversion matrix
		#subsequently, the rowsums are being calculated resulting in a vector of gamete ratios
		#"apply_along_axis(function, 2, 3Darray)" is similar to "apply(3D array, c(1,2), function)" in R
		gametes=np.apply_along_axis(ggr, 2, gametes)
#		print gametes[:,19:23,:]

		#back to genotypes. this where drift is acting, cf. gasa function
		intgen[1:(dimensions[0]+1), 1:(dimensions[1]+1), :]=np.apply_along_axis(gasa, 2, gametes)
		genot['c']+=1
		#print genot['c']
	genot['g']=intgen[1:(dimensions[0]+1), 1:(dimensions[1]+1), :]

	return genot
		
		


def __main__():

	#get global variable "selector"
	params()
	
	print 'Selection vector', selector


	genotypes=initialise()

	paramtab = open('sm%1.3f_dm%1.3f_sf%1.3f_df%1.3f_sh%1.3f_sd%1.3f_sc%1.3f_disp%1.3f_grad%1.3f_run%d_gens%05d.tab'%\
			(sm, dm, sf, df, sh, sd, sc, disp, grad, run, genotypes['c']), 'w')
	paramtab.write('Yint\tYslope\tYcent\tFint\tFslope\tFcent\n')

	for i in range(10):
		genotypes=tgb(genotypes, 10)


		Fh=Fhits(genotypes['g']).flatten()
		Fm=Fmisses(genotypes['g']).flatten()
		Yh=Yhits(genotypes['g']).flatten()
		Ym=Ymisses(genotypes['g']).flatten()

		yvals=stm.tools.add_constant(np.tile(np.arange(genotypes['g'].shape[1]),genotypes['g'].shape[0]))


		genotypes['modelparams']=[]
		glmy=stm.GLM(np.column_stack((Yh, Ym)), yvals, family=stm.families.Binomial())
		glmyp=glmy.fit().params
		genotypes['modelparams'] += [glmyp[0], glmyp[1], -glmyp[0]/glmyp[1]]

		glmf=stm.GLM(np.column_stack((Fh, Fm)), yvals, family=stm.families.Binomial())
		glmfp=glmf.fit().params
		genotypes['modelparams'] += [glmfp[0], glmfp[1], -glmfp[0]/glmfp[1]]

		pickle.dump(genotypes, gzip.open('sm%1.3f_dm%1.3f_sf%1.3f_df%1.3f_sh%1.3f_sd%1.3f_sc%1.3f_disp%1.3f_grad%1.3f_run%d_gens%05d.pyd.gz'%\
			(sm, dm, sf, df, sh, sd, sc, disp, grad, run, genotypes['c']), 'w'))


		out = gzip.open('sm%1.3f_dm%1.3f_sf%1.3f_df%1.3f_sh%1.3f_sd%1.3f_sc%1.3f_disp%1.3f_grad%1.3f_run%d_gens%05d.csv.gz'%\
			(sm, dm, sf, df, sh, sd, sc, disp, grad, run, genotypes['c']), 'w')		
		out.write(str(genotypes['g'].shape)[1:-1]+'\ny\tFhit\tFmis\tYhit\tYmis\n')
		np.savetxt(out, np.column_stack((yvals[:,1], Fh, Fm, Yh, Ym)), "%d", delimiter='\t')
		out.close()

		if genotypes['trans'][0]==0:
			if genotypes['modelparams'][2] < 0 or genotypes['modelparams'][2] > genotypes['g'].shape[1]:
				genotypes['trans'][0]=1

		if genotypes['trans'][1]==0:
			if genotypes['modelparams'][5] < 0 or genotypes['modelparams'][5] > genotypes['g'].shape[1]:
				genotypes['trans'][1]=1

		print "Generation %d, cline centres: %3.2f and %3.2f" % (genotypes['c'], genotypes['modelparams'][2], genotypes['modelparams'][5])
		paramtab.write('\t'.join(str(genotypes['modelparams'])[1:-1].split(', '))+'\n')		

		if 0 not in genotypes['trans']:
			print 'Both variants fixed. Break.'
			break

	paramtab.close()

## Plot using matplotlib
#	from mpl_toolkits.mplot3d import axes3d
#	import matplotlib.pyplot as plt
#	from matplotlib import cm
#
#	fig = plt.figure()
#	ax = fig.add_subplot(111, projection='3d')
#	Y = range(10)
#	X = range(90)
#	X, Y = np.meshgrid(X, Y)
#	Z = Fratio(genotypes['g'])
#	ax.plot_wireframe(X, Y, Z)
#	plt.show()



__main__()

